openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+	a.nodeListSynced = kubeInformers.Core().V1().Nodes().Informer().HasSynced
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nsListerSynced == nil {
+		return fmt.Errorf("%s plugin needs a namespace lister synced", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	if a.nodeListSynced == nil {
+		return fmt.Errorf("%s plugin needs a node lister synced", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	workloadAnnotations := getWorkloadAnnotations(pod.Annotations)
+
+	// strip any resource annotations specified by a user
+	stripResourcesAnnotations(pod.Annotations)
+	// strip any workload annotation to prevent from underlying components(CRI-O, kubelet) to apply any changes
+	// according to the workload annotation
+	stripWorkloadAnnotations(pod.Annotations)
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotations) > 1 {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningMultipleWorkloadAnnotations, workloadAnnotations)
+		return nil
+	}
+
+	workloadType, err := getWorkloadType(workloadAnnotations)
+	if err != nil {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningFailedGetWorkloadType, err)
+		return nil
+	}
+
+	// no workload annotation is available
+	if workloadType == "" {
+		return nil
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	// TODO: currently we do not check the effect, but it can be useful in the future
+	_, err = getWorkloadAnnotationEffect(workloadAnnotations[workloadAnnotation])
+	if err != nil {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningFailedGetWorkloadAnnotationEffect, err)
+		return nil
+	}
+
+	if !a.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(attr, fmt.Errorf("%s node and namespace caches not synchronized", PluginName))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
+		}
+
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if !isNamespaceAllowed(ns.Annotations, workloadType) {
+		return nil
```

Comment:
What is the use-case for allowing a pod to be created that doesn't have its intent honored?

Reply:
If the pod has requested a workload type and the namespace does not allow it, we should reject the pod.  Our namespaces can unconditionally set the annotation to allow these types so even if the feature is not turned on, the annotation can be set.